src/lib/Types/Core.hs

```diff
@@ -131,22 +131,17 @@ data FieldRowElem (n::S)
 data DataDef n where
   -- The `SourceName` is just for pretty-printing. The actual alpha-renamable
   -- binder name is in UExpr and Env
-  DataDef :: SourceName -> DataDefBinders n l -> [DataConDef l] -> DataDef n
+  DataDef :: SourceName -> Nest PiBinder n l -> [DataConDef l] -> DataDef n
 
 data DataConDef n =
   -- Name for pretty printing, constructor elements, representation type,
   -- list of projection indices that recovers elements from the representation.
   DataConDef SourceName (Type n) [[Projection]]
   deriving (Show, Generic)
 
-data DataDefBinders n l where
-  DataDefBinders
-    :: Nest Binder n h  -- ordinary params
-    -> Nest Binder h l  -- dict params
-    -> DataDefBinders n l
-
-data DataDefParams n = DataDefParams [Atom n] [Atom n]  -- ordinary params, dict params
-                       deriving (Show, Generic)
+-- Left is ordinary binders, Right is class binders
+newtype DataDefParams n = DataDefParams [EitherE Atom Atom n]
```

Comment:
This `EitherE` is a good candidate for a new ADT.

Reply:
Replacing with [(Arrow, Atom n)]